packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -77,179 +81,220 @@ bool _shouldRunPubGet({ File pubSpecYaml, File dotPackages }) {
   return false;
 }
 
-/// [context] provides extra information to package server requests to
-/// understand usage.
-Future<void> pubGet({
-  @required PubContext context,
-  String directory,
-  bool skipIfAbsent = false,
-  bool upgrade = false,
-  bool offline = false,
-  bool checkLastModified = true,
-  bool skipPubspecYamlCheck = false,
-}) async {
-  directory ??= fs.currentDirectory.path;
-
-  final File pubSpecYaml = fs.file(fs.path.join(directory, 'pubspec.yaml'));
-  final File dotPackages = fs.file(fs.path.join(directory, '.packages'));
-
-  if (!skipPubspecYamlCheck && !pubSpecYaml.existsSync()) {
-    if (!skipIfAbsent) {
-      throwToolExit('$directory: no pubspec.yaml found');
+/// A handle for interacting with the pub tool.
+abstract class Pub {
+  /// Create a default [Pub] instance.
+  const factory Pub() = _DefaultPub;
+
+  /// [context] provides extra information to package server requests to
+  /// understand usage.
+  Future<void> pubGet({
+    @required PubContext context,
+    String directory,
+    bool skipIfAbsent = false,
+    bool upgrade = false,
+    bool offline = false,
+    bool checkLastModified = true,
+    bool skipPubspecYamlCheck = false,
+  });
+
+  /// Runs pub in 'batch' mode, forwarding complete lines written by pub to its
+  /// stdout/stderr streams to the corresponding stream of this process, optionally
+  /// applying filtering. The pub process will not receive anything on its stdin stream.
+  ///
+  /// The `--trace` argument is passed to `pub` (by mutating the provided
+  /// `arguments` list) when `showTraceForErrors` is true, and when `showTraceForErrors`
+  /// is null/unset, and `isRunningOnBot` is true.
+  ///
+  /// [context] provides extra information to package server requests to
+  /// understand usage.
+  Future<void> pub(
```

Comment:
Going by the comment, maybe this can be renamed 'batch' to avoid `pub.pub(...)`

Reply:
Renamed this batch and the other to interactively.